teps/0022-trigger-immutable-input.md

```diff
@@ -0,0 +1,351 @@
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+- [Upgrade and Migration Strategy](#upgrade-and-migration-strategy)
+  <!-- /toc -->
+
+## Summary and Motivation
+
+<!--
+This section is for explicitly listing the motivation, goals and non-goals of
+this TEP.  Describe why the change is important and the benefits to users.  The
+motivation section can optionally provide links to [experience reports][] to
+demonstrate the interest in a TEP within the wider Tekton community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+Today with overlays, users have the ability to modify incoming event payloads
+either with CEL or Webhook interceptors.
+
+For example, the following
+[CEL Interceptor](https://github.com/tektoncd/triggers/blob/master/docs/eventlisteners.md#CEL-Interceptors)
+can modify the incoming event payload to include a shortened SHA:
+
+```yaml
+- key: body.pull_request.head.short_sha
+  expression: "truncate(body.pull_request.head.sha, 7)"
+```
+
+[Webhook Interceptors](https://github.com/tektoncd/triggers/blob/master/docs/eventlisteners.md#Webhook-Interceptors)
+can provide similar functionality with their own custom code, but they are also
+required to pass back bodies in responses if they wish to preserve data.
+
+While these features enable users to include missing data required for trigger
+execution, this has a few challenges:
+
+- Auditability - looking ahead to storing Trigger event data as part of
+  [long term results](https://github.com/tektoncd/community/pull/217), modifying
+  the event payload makes it difficult to look back and verify the incoming
+  event. Was a value provided by GitHub, or an interceptor? What happens if the
+  input no longer matches any included signatures?
+- It's easy for webhook interceptors to accidentally modify the input payload
+  and cause issues for downstream interceptors, potentially making it difficult
+  for interceptors to be composable.
+
+## Requirements
+
+<!--
+List the requirements for this TEP.
+-->
+
+- Ensure input events can be passed through during trigger processing
+  unmodified.
+- Provide backwards compatibility for at least 1 release.
+
+## Proposal and Design Details
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation.  The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+### Webhook Interceptors
+
+#### Input Event Field
+
+Instead of mixing input event fields with others (e.g. the `webhook.header`
+field) we will include the entire input event in a reserved body field named
+`input`. This field will not be mutated throughout trigger processing, so that
+inteceptors and trigger bindings can rely on assumptions on the incoming event
+type. Incoming event data will be stored within `input` as `body` and `headers`
```

Comment:
Looking at it again, maybe the interface becomes:
```go
interface Interceptor {
    ExecuteTrigger(headers http.Header, body map[string]interface{}) (extensions map[string]interface{}, err error)
}
```
The Webhook interceptor would just do a POST, transferring the headers, encoding the body, and being expected to return an optional JSON body (which would become extensions) ?

Reply:
I think we'll want to encapsulate this into Request/Response types, following what is proposed in the [pluggable interceptor proposal](https://docs.google.com/document/d/1zIG295nyWonCXhb8XcOC41q4YVP9o2Hgg8uAoL4NdPA/edit#heading=h.ecpe4nobzdj5) by @dibyom. This will make it easier to add additional fields into the request/response objects as needed (i.e. the additional trigger fields proposed by Dibyo):

```go
interface Interceptor {
    ExecuteTrigger(req InterceptorRequest) (InterceptorResponse, error)
}

type InterceptorRequest struct {
  input HTTPEvent
  interceptors map[string]interface{}
}

type HTTPEvent struct {
  Headers http.Headers
  Body json.RawMessage
}

type InterceptorResponse struct {
  Body json.RawMessage
}
```